<div align="right" style=" font-size: 80%; text-align: center; margin: 0 auto">
<img
 src="https://raw.githubusercontent.com/Explore-AI/Pictures/master/alx-courses/aice/assets/Content_page_banner_blue_dots.png"
 alt="ALX Content Header"
 class="full-width-image"
/>
</div>

# Filtering and analysing a summary statistic report

In this notebook, we demonstrate how to filter and analyse a summary statistic report using the `WHERE` and `HAVING` clauses.

> ⚠️ This notebook will not run on Google Colab because it cannot connect to a local database. Please make sure that this notebook is running on the same local machine as your MySQL Workbench installation and MySQL `united_nations` database.

## Learning objectives

By the end of this train, you should:
-  Know how to analyse a summary statistic report using the `WHERE` and `HAVING` clauses.

## Connecting to our MySQL database


Continuing with the numerical analysis of our Access_to_Basic_Services table created in MySQL Workbench, we want to filter and analyse our already-created summary statistic report. We can apply the same queries in MySQL Workbench and in this notebook if we connect to our MySQL server. Since we have a MySQL database, we can connect to it using mysql and pymysql.

In [1]:
!pip install --upgrade ipython-sql sqlalchemy pymysql prettytable

  Using cached prettytable-3.16.0-py3-none-any.whl.metadata (33 kB)
Using cached prettytable-3.16.0-py3-none-any.whl (33 kB)
  Attempting uninstall: prettytable
    Found existing installation: prettytable 3.8.0
    Uninstalling prettytable-3.8.0:
      Successfully uninstalled prettytable-3.8.0


In [1]:
!pip install "prettytable==3.8.0"

  Using cached prettytable-3.8.0-py3-none-any.whl.metadata (26 kB)
Using cached prettytable-3.8.0-py3-none-any.whl (27 kB)
  Attempting uninstall: prettytable
    Found existing installation: prettytable 3.16.0
    Uninstalling prettytable-3.16.0:
      Successfully uninstalled prettytable-3.16.0


In [1]:
# Import necessary Python libraries
import os                     # Used to work with environment variables and system paths
from getpass import getpass   # Used to securely input passwords (hidden while typing)
from urllib.parse import quote_plus  # Used to safely encode special characters in the password

# Prompt the user to enter their MySQL password securely
pw = getpass("Enter MySQL password: ")    

# Encode the password to handle special characters (e.g., @, #, $, etc.)
pw_enc = quote_plus(pw)                   

# Create a database connection string and store it as an environment variable
# This string will be used by SQLAlchemy or ipython-sql to connect to MySQL
os.environ["DATABASE_URL"] = f"mysql+pymysql://root:{pw_enc}@localhost:3306/united_nations"

# Print a confirmation message indicating that the environment variable has been set
print("DATABASE_URL set for this session.")

Enter MySQL password:  ········


DATABASE_URL set for this session.


In [2]:
# Load the SQL extension for Jupyter Notebook.
# This enables the use of %sql and %%sql magic commands to run SQL queries directly in notebook cells.
%load_ext sql

# Establish a connection to the database using the connection string stored in the environment variable DATABASE_URL.
# This connects Jupyter to your MySQL database, allowing you to execute SQL commands interactively.
%sql $DATABASE_URL

To make a query, we add the `%%sql` command to the start of a cell, create one open line, and then the query like below, and run the cell.

In [3]:
%%sql

SELECT 
    *
FROM
    Access_to_Basic_Services
LIMIT 5;

 * mysql+pymysql://root:***@localhost:3306/united_nations
5 rows affected.


Region,Sub_region,Country_name,Time_period,Pct_managed_drinking_water_services,Pct_managed_sanitation_services,Est_population_in_millions,Est_gdp_in_billions,Land_area,Pct_unemployment
Central and Southern Asia,Central Asia,Kazakhstan,2015,94.67,98.00,17.542806,184.39,2699700.00,4.93
Central and Southern Asia,Central Asia,Kazakhstan,2016,94.67,98.00,17.794055,137.28,2699700.00,4.96
Central and Southern Asia,Central Asia,Kazakhstan,2017,95.00,98.00,18.037776,166.81,2699700.00,4.90
Central and Southern Asia,Central Asia,Kazakhstan,2018,95.00,98.00,18.276452,179.34,2699700.00,4.85
Central and Southern Asia,Central Asia,Kazakhstan,2019,95.00,98.00,18.513673,181.67,2699700.00,4.80


## Exercise

We started with finding out the minimum, maximum, and average percentage of people that have access to drinking water services, the number of countries, and the total GDP per region and sub-region. We also ordered this data by estimated GDP.

In [4]:
%%sql

SELECT Region, 
    Sub_region,
    MIN(Pct_managed_drinking_water_services) AS min_Pct_managed_drinking_water_services, 
    MAX(Pct_managed_drinking_water_services) AS max_Pct_managed_drinking_water_services, 
    AVG(Pct_managed_drinking_water_services) AS avg_Pct_managed_drinking_water_services,
    COUNT(DISTINCT(Country_name)) AS Number_of_countries,
    SUM(EST_gdp_in_billions) AS EST_total_gdp_in_billions 
FROM united_nations.Access_to_Basic_Services 
GROUP BY Region, Sub_region
ORDER BY EST_total_gdp_in_billions;

 * mysql+pymysql://root:***@localhost:3306/united_nations
18 rows affected.


Region,Sub_region,min_Pct_managed_drinking_water_services,max_Pct_managed_drinking_water_services,avg_Pct_managed_drinking_water_services,Number_of_countries,EST_total_gdp_in_billions
Oceania,Polynesia,91.00,100.00,98.506481,9,46.76
Oceania,Micronesia,73.67,100.00,93.414634,7,46.92
Oceania,Melanesia,53.67,99.00,82.700333,5,237.87
Sub-Saharan Africa,Middle Africa,38.33,77.33,58.967059,9,837.34
Central and Southern Asia,Central Asia,80.33,100.00,93.144667,5,1670.32
Sub-Saharan Africa,Eastern Africa,46.33,100.00,70.128421,20,1944.61
Latin America and the Caribbean,Caribbean,64.00,100.00,96.005000,27,2070.17
Sub-Saharan Africa,Southern Africa,73.33,92.00,82.645000,5,2386.64
Northern Africa and Western Asia,Northern Africa,61.33,100.00,88.906111,6,2736.80
Sub-Saharan Africa,Western Africa,53.33,99.00,72.365686,17,3621.31


Let's continue with our analysis.

We want to use the summary statistic report to do the following:

1. Filter for the year 2020.
2. Focus on countries where the percentage of managed drinking water services is below 60%.
3. Filter for the regions and sub-regions that have fewer than four countries.

### 1. Filter for the year 2020.
Using the above query, focus on results where the time period is 2020 using the `WHERE` clause.

In [6]:
%%sql

SELECT 
    Region, 
    Sub_region,
    Time_period,
    MIN(Pct_managed_drinking_water_services) AS min_Pct_managed_drinking_water_services, 
    MAX(Pct_managed_drinking_water_services) AS max_Pct_managed_drinking_water_services, 
    AVG(Pct_managed_drinking_water_services) AS avg_Pct_managed_drinking_water_services,
    COUNT(DISTINCT(Country_name)) AS Number_of_countries,
    SUM(EST_gdp_in_billions) AS EST_total_gdp_in_billions 
FROM 
    united_nations.Access_to_Basic_Services 
Where
    Time_period = 2020
GROUP BY 
    Region, Sub_region
ORDER BY 
    EST_total_gdp_in_billions
;

 * mysql+pymysql://root:***@localhost:3306/united_nations
18 rows affected.


Region,Sub_region,Time_period,min_Pct_managed_drinking_water_services,max_Pct_managed_drinking_water_services,avg_Pct_managed_drinking_water_services,Number_of_countries,EST_total_gdp_in_billions
Oceania,Micronesia,2020,77.00,100.00,94.500000,6,6.67
Oceania,Polynesia,2020,92.00,100.00,98.555556,9,7.84
Oceania,Melanesia,2020,56.67,99.00,82.934000,5,40.21
Sub-Saharan Africa,Middle Africa,2020,38.33,77.33,59.332500,8,123.22
Central and Southern Asia,Central Asia,2020,85.00,100.00,94.134000,5,239.10
Latin America and the Caribbean,Caribbean,2020,65.00,100.00,95.910667,15,343.26
Sub-Saharan Africa,Eastern Africa,2020,48.33,100.00,70.018824,17,359.10
Sub-Saharan Africa,Southern Africa,2020,76.33,92.00,83.668000,5,369.34
Northern Africa and Western Asia,Northern Africa,2020,62.33,100.00,90.053333,6,386.29
Sub-Saharan Africa,Western Africa,2020,53.33,99.00,73.607059,17,631.91


### 2. Focus on countries where the percentage of managed drinking water services is below 60%.

Adding onto your query above, focus on results where the percentage of managed drinking water services is smaller than 60% using the `WHERE` clause.

In [7]:
%%sql

SELECT 
    Region, 
    Sub_region,
    Time_period,
    MIN(Pct_managed_drinking_water_services) AS min_Pct_managed_drinking_water_services, 
    MAX(Pct_managed_drinking_water_services) AS max_Pct_managed_drinking_water_services, 
    AVG(Pct_managed_drinking_water_services) AS avg_Pct_managed_drinking_water_services,
    COUNT(DISTINCT(Country_name)) AS Number_of_countries,
    SUM(EST_gdp_in_billions) AS EST_total_gdp_in_billions 
FROM 
    united_nations.Access_to_Basic_Services 
Where
    Time_period = 2020
    And
        Pct_managed_drinking_water_services < 60
GROUP BY 
    Region, Sub_region
ORDER BY 
    EST_total_gdp_in_billions
;

 * mysql+pymysql://root:***@localhost:3306/united_nations
4 rows affected.


Region,Sub_region,Time_period,min_Pct_managed_drinking_water_services,max_Pct_managed_drinking_water_services,avg_Pct_managed_drinking_water_services,Number_of_countries,EST_total_gdp_in_billions
Oceania,Melanesia,2020,56.67,56.67,56.670000,1,23.85
Sub-Saharan Africa,Western Africa,2020,53.33,57.33,55.330000,2,31.67
Sub-Saharan Africa,Middle Africa,2020,38.33,52.67,47.750000,4,66.67
Sub-Saharan Africa,Eastern Africa,2020,48.33,58.00,54.997500,4,127.59


### 3. Filter for the sub-regions that have fewer than four countries.

Filter the results above to only include the regions and sub-regions that have fewer than four countries in the Number_of_countries alias using the `HAVING` clause.

In [9]:
%%sql

SELECT 
    Region, 
    Sub_region,
    Time_period,
    MIN(Pct_managed_drinking_water_services) AS min_Pct_managed_drinking_water_services, 
    MAX(Pct_managed_drinking_water_services) AS max_Pct_managed_drinking_water_services, 
    AVG(Pct_managed_drinking_water_services) AS avg_Pct_managed_drinking_water_services,
    COUNT(DISTINCT(Country_name)) AS Number_of_countries,
    SUM(EST_gdp_in_billions) AS EST_total_gdp_in_billions 
FROM 
    united_nations.Access_to_Basic_Services 
Where
    Time_period = 2020
    And
        Pct_managed_drinking_water_services < 60
GROUP BY 
    Region, Sub_region
Having
    Number_of_countries < 4
ORDER BY 
    EST_total_gdp_in_billions ASC
;

 * mysql+pymysql://root:***@localhost:3306/united_nations
2 rows affected.


Region,Sub_region,Time_period,min_Pct_managed_drinking_water_services,max_Pct_managed_drinking_water_services,avg_Pct_managed_drinking_water_services,Number_of_countries,EST_total_gdp_in_billions
Oceania,Melanesia,2020,56.67,56.67,56.670000,1,23.85
Sub-Saharan Africa,Western Africa,2020,53.33,57.33,55.330000,2,31.67


## Solutions

### 1. Filter for the year 2020.

In [ ]:
%%sql

SELECT Region, 
    Sub_region,
    MIN(Pct_managed_drinking_water_services) AS min_Pct_managed_drinking_water_services, 
    MAX(Pct_managed_drinking_water_services) AS max_Pct_managed_drinking_water_services, 
    AVG(Pct_managed_drinking_water_services) AS avg_Pct_managed_drinking_water_services, 
    COUNT(DISTINCT(Country_name)) AS Number_of_countries,
    SUM(EST_gdp_in_billions) AS EST_total_gdp_in_billions
FROM Access_to_Basic_Services 
WHERE Time_period = 2020
GROUP BY Region, Sub_region
ORDER BY EST_total_gdp_in_billions ASC;


### 2. Focus on countries where the percentage of managed drinking water services is below 60%.

In [ ]:
%%sql

SELECT Region, 
    Sub_region,
    MIN(Pct_managed_drinking_water_services) AS min_Pct_managed_drinking_water_services, 
    MAX(Pct_managed_drinking_water_services) AS max_Pct_managed_drinking_water_services, 
    AVG(Pct_managed_drinking_water_services) AS avg_Pct_managed_drinking_water_services, 
    COUNT(DISTINCT(Country_name)) AS Number_of_countries,
    SUM(EST_gdp_in_billions) AS EST_total_gdp_in_billions
FROM Access_to_Basic_Services 
WHERE Time_period = 2020
    AND Pct_managed_drinking_water_services < 60
GROUP BY Region, Sub_region
ORDER BY EST_total_gdp_in_billions ASC;

### 3. Filter for the regions and sub-regions that have fewer than four countries.

In [ ]:
%%sql

SELECT Region, 
    Sub_region,
    MIN(Pct_managed_drinking_water_services) AS min_Pct_managed_drinking_water_services, 
    MAX(Pct_managed_drinking_water_services) AS max_Pct_managed_drinking_water_services, 
    AVG(Pct_managed_drinking_water_services) AS avg_Pct_managed_drinking_water_services, 
    COUNT(DISTINCT(Country_name)) AS Number_of_countries,
    SUM(EST_gdp_in_billions) AS EST_total_gdp_in_billions
FROM Access_to_Basic_Services 
WHERE Time_period = 2020
    AND Pct_managed_drinking_water_services < 60
GROUP BY Region, Sub_region
HAVING Number_of_countries < 4
ORDER BY EST_total_gdp_in_billions ASC;

The `WHERE` clause may come to mind first when trying to apply this criterion. However, take note that this criterion may only be used after the "Number_of_countries" aggregate and grouping. This is because we want to group the data by region and sub-region and then only choose the groups that have fewer than four countries inside those groups.
Because the `WHERE` clause executes before the aggregate and GROUP BY clauses, we are unable to use it.
Therefore the `HAVING` clause is more appropriate here.

With this report, we can answer questions like “Out of the sub-regions that meet the criteria, which sub-region has the lowest GDP?”
Since the data is ordered according to the estimated GDP in ascending order, the first result in the sub-region column will contain the answer. That is, Melanesia at a GDP of 23.85 billion.


<div align="center" style=" font-size: 80%; text-align: center; margin: 0 auto">
<img src="https://raw.githubusercontent.com/Explore-AI/Pictures/refs/heads/master/ALX_banners/ALX_Navy.png"  style="width:100px"  ;/>
</div>